substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -672,50 +674,41 @@ impl<T: Config> Pallet<T> {
 		let mut source_delegation =
 			Delegators::<T>::get(&source_delegator).defensive_ok_or(Error::<T>::BadState)?;
 
-		// some checks that must have already been checked before.
+		// ensure source has enough funds to migrate.
 		ensure!(source_delegation.amount >= amount, Error::<T>::NotEnoughFunds);
 		debug_assert!(
 			!Self::is_delegator(&destination_delegator) && !Self::is_agent(&destination_delegator)
 		);
 
 		let agent = source_delegation.agent.clone();
-		// update delegations
-		Delegation::<T>::new(&agent, amount).update_or_kill(&destination_delegator);
+		// create a new delegation for destination delegator.
+		let _ = Delegation::<T>::new(&agent, amount).update(&destination_delegator);
+		// Provide for the delegator account. This ensures that amount including existential deposit
+		// can be held.
+		frame_system::Pallet::<T>::inc_providers(&destination_delegator);
 
 		source_delegation.amount = source_delegation
 			.amount
 			.checked_sub(&amount)
 			.defensive_ok_or(Error::<T>::BadState)?;
 
-		source_delegation.update_or_kill(&source_delegator);
-
-		// release funds from source
-		let released = T::Currency::release(
+		// transfer the held amount in `source_delegator` to `destination_delegator`.
+		let _ = T::Currency::transfer_on_hold(
 			&HoldReason::StakingDelegation.into(),
-			&source_delegator,
-			amount,
-			Precision::BestEffort,
-		)?;
-
-		defensive_assert!(released == amount, "hold should have been released fully");
-
-		// transfer the released amount to `destination_delegator`.
-		let post_balance = T::Currency::transfer(
 			&source_delegator,
 			&destination_delegator,
 			amount,
-			Preservation::Expendable,
-		)
-		.map_err(|_| Error::<T>::BadState)?;
+			Precision::Exact,
+			Restriction::OnHold,
+			Fortitude::Polite,
+		)?;
 
-		// if balance is zero, clear provider for source (proxy) delegator.
-		if post_balance == Zero::zero() {
+		// update source delegation.
+		if source_delegation.update(&source_delegator) {
```

Comment:
same here as above -- the decreasing providers could be called within `fn update` if relevant.